src/panel/tag_parser/parser.js

```diff
@@ -0,0 +1,206 @@
+	}
+
+	let tokenArr = [];
+	let i = 0;
+	let char = "", expr = "";
+
+	while (i < string.length) {
+		char = string.charAt(i);
+
+		// If char is whitespace
+		if (/\s/.test(char)) {
+			i += 1;
+		}
+		else if (char in tokenDict) {
+			tokenArr.push(tokenDict[char]);
+			i += 1;
+		}
+		else {
+			expr = "";
+
+			while (!(char in tokenDict) && i < string.length) {
+				expr += char;
+				
+				i += 1;
+				char = string.charAt(i);
+			}
+
+			tokenArr.push({
+				tokenType: TokenType.LITERAL, 
+				value: expr.trim(), 
+				precedence: 9
+			});
+		}
+	}
+
+	console.log("Infix tokenArr on next line");
+	prettyPrintTokenArr(tokenArr);
+
+	return tokenArr;
+}
+
+function isInputValid(string) {
+	const input = string.trim();
+	const firstChar = input.charAt(0), lastChar = input.charAt(input.length - 1);
+
+	if (firstChar === "&" || firstChar === "|" || firstChar === ",") {
+		return false;
+	}
+	if (lastChar === "&" || lastChar === "|" || lastChar === ",") {
+		return false;
+	}
+
+	return true;
+}
+
+// Shunting yard algorithm
+function convertTokenArrToPostfix(tokenArr) {
+	let postfixArr = [];
+	let stack = [];
+
+	for (let i=0; i<tokenArr.length; i++) {
+		switch (tokenArr[i]["tokenType"]) {
+		
+		case TokenType.LITERAL:
+			postfixArr.push(tokenArr[i]);
+			break;
+
+		case TokenType.BINARY_OP:
+			// If stack is empty or contains left parenthesis
+			if (stack.length === 0 || peek(stack)["tokenType"] === TokenType.OPEN_PAREN) {
+				stack.push(tokenArr[i]);
+			}
+			// If operator has higher precedence than the one on the stack
+			else if (tokenArr[i]["precedence"] <= peek(stack)["precedence"]) {
+				stack.push(tokenArr[i]);
+			}
+			// If operator has lower precedence than the one on the stack
+			else {
+				while (tokenArr[i]["precedence"] > peek(stack)["precedence"]) {
+					postfixArr.push(stack.pop());
+				}
+			}
+			break;
+
+		case TokenType.OPEN_PAREN:
+			stack.push(tokenArr[i]);
+			break;
+
+		case TokenType.CLOSE_PAREN:
+			if (stack.length === 0) {
+				throw new Error("There are unbalanced brackets in the expression.");
+			}
+
+			while (peek(stack)["tokenType"] !== TokenType.OPEN_PAREN) {
+				postfixArr.push(stack.pop());
+			}
+			// Discard open paren
+			stack.pop();
+			break;
+		}
+	}
+
+	while (stack.length !== 0) {
+		const token = stack.pop();
+
+		if (token["tokenType"] === TokenType.OPEN_PAREN) {
+			throw new Error("There are unbalanced brackets in the expression.");
+		}
+
+		postfixArr.push(token);
+	}
+
+	console.log("Postfix tokenArr on next line");
+	prettyPrintTokenArr(postfixArr);
+
+	if (!isPostfixArrValid(postfixArr)) {
+		throw new Error("Unable to parse tags input.");
+	}
+
+	return postfixArr;
+}
+
+function peek(stack) {
+	return stack[stack.length - 1];
+}
+
+function isPostfixArrValid(arr) {
+	let stackCount = 0;
+
+	for (let i=0; i<arr.length; i++) {
+		if (arr[i]["tokenType"] === TokenType.LITERAL) {
+			stackCount += 1;
+		}
+		// If token is operator
+		else {
+			stackCount -= 1;
+		}

```

Comment:
Maybe a check to see whether when an operator is encountered, at least 2 items are on the stack? 😄